ams/optimizer/spark-optimizer/src/main/java/com/netease/arctic/optimizer/spark/SparkOptimizer.java

```diff
@@ -0,0 +1,161 @@
+
+  public static void main(String[] args) throws Exception {
+    SparkSession spark = SparkSession.builder().appName(APP_NAME).getOrCreate();
+    JavaSparkContext jsc = new JavaSparkContext(spark.sparkContext());
+    SparkOptimizerConfig config = new SparkOptimizerConfig(args);
+    if (!jsc.getConf().getBoolean("spark.dynamicAllocation.enabled", false)) {
+      LOG.warn(
+          "To better utilize computing resources, it is recommended to enable 'spark.dynamicAllocation.enabled' "
+              + "and set 'spark.executor.cores' * 'spark.dynamicAllocation.maxExecutors' greater than or equal to 'OPTIMIZER_EXECUTION_PARALLEL'");
+    }
+    int driverMemory = Utils.memoryStringToMb(jsc.getConf().get("spark.driver.memory", "1g"));
+    if (config.getMemorySize() < driverMemory) {
+      config.setMemorySize(driverMemory);
+    }
+
+    OptimizerToucher toucher = new OptimizerToucher(config);
+    if (config.getResourceId() != null) {
+      toucher.withRegisterProperty(PropertyNames.RESOURCE_ID, config.getResourceId());
+    }
+    toucher.withRegisterProperty(Resource.PROPERTY_JOB_ID, spark.sparkContext().applicationId());
+    LOG.info("Starting optimizer with configuration:{}", config);
+
+    int threadId = Thread.currentThread().hashCode();
+    SparkExecutor sparkExecutor = new SparkExecutor(jsc, config, threadId);
+    ThreadFactory consumerFactory =
+        new ThreadFactoryBuilder()
+            .setDaemon(false)
+            .setNameFormat("spark-optimizing-puller-%d")
+            .build();
+    ExecutorService pullerService = Executors.newSingleThreadExecutor(consumerFactory);
+    TaskPuller puller = new TaskPuller(config, threadId, sparkExecutor);
+    pullerService.execute(puller);
+    toucher
+        .withTokenChangeListener(
+            newToken -> {
+              puller.setToken(newToken);
+              sparkExecutor.setToken(newToken);
+            })
+        .start();
+  }
+
+  static class TaskPuller extends AbstractOptimizerOperator implements Runnable {
+    private final SparkExecutor sparkExecutor;
+    private final SparkOptimizerConfig config;
+    private final int threadId;
+
+    private TaskPuller(SparkOptimizerConfig config, int threadId, SparkExecutor sparkExecutor) {
+      super(config);
+      this.config = config;
+      this.threadId = threadId;
+      this.sparkExecutor = sparkExecutor;
+    }
+
+    @Override
+    public void run() {
+      try {
+        ThreadPoolExecutor executorService = sparkExecutor.getExecutorService();
+        LOG.info("Starting to poll optimizing tasks.");
+        while (true) {
+          int activeCount = executorService.getActiveCount();
+          if (activeCount >= executorService.getMaximumPoolSize()) {
+            LOG.info("Polled {} tasks, the queue is full", activeCount);
+            waitAShortTime();
+          } else {
+            OptimizingTask task = pollTask();
+            if (task != null && ackTask(task)) {
```

Comment:
If I understand correctly, the goal of the Spark optimizer is to leverage batch processing. So we may need to fetch a batch of tasks at each optimization?
WDYT @shidayang , @zhoujinsong

Reply:
I don't agree. I believe that compared to the Flink Optimizer, the Spark Optimizer is more flexible in actual resource utilization. It can adjust the number of executors based on the actual load of the Optimizing-Group. 

When the spark executer is idle, it can automatically reduce the number of executor nodes. In comparison, under the Flink Optimizer, concurrency strictly equals resource usage, while under the Spark Optimizer, concurrency represents the upper limit of resource usage.

If we adopt the approach of acquiring a batch of tasks before execution, we cannot guarantee the timeliness of optimization on the table. 

In the current design, plans are triggered based on statistics of small files on the table, which means that the tasks generated by the plan must be executed as soon as possible.